src/org/etools/j1939_84/controllers/part12/Part12Step12Controller.java

```diff
@@ -0,0 +1,248 @@
+                                                                         addFailure("6.12.12.c - Unused bytes in DM58 are not padded with FFh in the response from "
+                                                                                 + moduleName);
+                                                                     }
+                                                                     // d. Fail, if data returned is greater than FBh
+                                                                     // (for 1 byte SP), FBFFh (for 2 byte SP), or
+                                                                     // FBFFFFFFh (for 4 byte SP)
+                                                                     if (isGreaterThanFb(packet)) {
+                                                                         addFailure("6.12.12.d - Data returned is greater than 0xFB... threshold");
+                                                                     }
+                                                                 }
+                                                             })
+                                                             .collect(Collectors.toList());
+
+                      // b. Fail, if DM58 not received (after allowed retries)
+                      if (packets.isEmpty()) {
+                          addFailure("6.12.12.d - DM58 not received (after allowed retries)");
+                      }
+
+                  });
+        });
+
+        // Actions3:
+        // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
+        // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
+        // preferred when available).
+        // Fail/Warn criteria3:
+        // a. Fail if a NACK is not received
+        obdModules.forEach(module -> {
+            getDM58(module.getSourceAddress());
+        });
+    }
+
+    private void getDM58(int moduleAddress) {
+        var allSps = getDataRepository().getObdModules()
+                                        .stream()
+                                        .flatMap(m -> m.getFilteredDataStreamSPNs().stream())
+                                        .collect(Collectors.toList());
+        SupportedSPN requestSpn = null;
+
+        if (allSps.isEmpty()) {
+            addWarning("6.12.12.3 - No modules reported supported SPs");
+        } else {
+            requestSpn = allSps.get(0);
+        }
+
+        var nonRatFaultSps = allSps.stream()
+                                   .filter(spn -> !spn.supportsRationalityFaultData())
+                                   .collect(Collectors.toList());
+
+        if (nonRatFaultSps.isEmpty()) {
+            addWarning("6.12.12.3.a - No SPs found that do NOT indicate support for DM58 in the DM24 response - using first SP found");
```

Comment:
Per our standards, the text from the document should be near the code that’s fulfilling the requirement.  I don’t see the text for this failure (or others)